adoc/chapters/architecture.adoc

```diff
@@ -1778,8 +1778,13 @@ certain arguments to a <<sycl-kernel-function>> (as described in
 <<device-copyable>> as defined below.
 
 Any type that is trivially copyable (as defined by the {cpp} core language) is
-implicitly device copyable.  In addition, the following types from the {cpp}
-core language are implicitly device copyable:
+implicitly device copyable.
+
+Although implementations are not required to support device code that calls
+library functions from the {cpp} core language, some implementations may
+provide device support for some of these functions.  If the implementation
```

Comment:
I wonder: how is user supposed to check if implementation supports a certain `std::` container? Is it responsibility of every implementation to document what's exactly is supported? 

I'm asking, because we have SYCL-CTS which checks that types listed below are device copyable:

https://github.com/KhronosGroup/SYCL-CTS/blob/4a9e68f81e4c63ace2c50ef1c68068b329ab9573/tests/common/type_coverage.h#L541

With the new wording, it seems like implementation is free to **not** mark them as device copyable if they are not supported in SYCL kernels. Should we remove the test then, because specification now provides no guarantees here?

Reply:
The Khronos SYCL WG should discuss if this PR is what we intended.  (I think it is.)  Answering your questions assuming this PR is accepted:

> I wonder: how is user supposed to check if implementation supports a certain std:: container? Is it responsibility of every implementation to document what's exactly is supported?

Yes.  The core SYCL spec does not guarantee that any `std` functions or types are available in device code.  Of course, an implementation is allowed to support some of these functions as an  extension if it wants to, though.  If an application wants to be portable across SYCL implementations, it needs to use `#ifdef` around any use of `std` functions to make sure they are not used unless the implementation supports this extension.

> With the new wording, it seems like implementation is free to not mark them as device copyable if they are not supported in SYCL kernels. Should we remove the test then, because specification now provides no guarantees here?

Yes, I think we should remove this part of the test.

Reply:
There are already some `std::`function usages in the CTS: https://github.com/KhronosGroup/SYCL-CTS/pull/505/files#diff-9f869743b3173b1554dab7dcd20be5a91239ec33aff3ef8cec9ad2aeac5dadf0R64

Reply:
I have seen recently some `std::memcmp` into some CTS device code.

Reply:
It would be great if we could come to an agreement about this PR soon.  Some of the CTS work for SYCL 2020 is being held up because of uncertainty about whether these `std` types need to be supported in device code.  If we accept this PR, then we would change the CTS to remove use of these `std` types in device code.

I've added the "P1" label to the associated internal issue 533 in order to raise visibility for the next WG meeting.

Reply:
@gmlueck As of C++23, both `span` and `basic_string_view` are guaranteed to be trivially copyable, and no major compiler vendor expects that their current implementation will change.  `mdspan` is also specified to be trivially copyable if its component parts are trivially copyable, and all the ones provided by the standard are trivially copyable.

That leaves `array`, which may already be required to be trivially copyable if either `T` is trivially copyable or `N == 0` (because it has to be an aggregate) even though the standard doesn't explicitly say so.

Reply:
Are you suggesting that we remove these types from the list because they are already trivially copyable, which means they are also device copyable?

I am not opposed to doing this.  My main goal in this PR to is to clarify that there is no requirement for an implement to support these types in device code.

Reply:
@gmlueck I'm fine with removing them.

I think there is a deeper question here.  If a standard library implementation of a type is trivially copyable, how are users to know whether or not that type is supported in device code?

Making things more complicated:  while `array` of a trivially copyable `T` or `N==0` is trivially copyable, it contains operations that cannot be supported in device code.  An example is `array::at()`, because it throws exceptions.

Reply:
> I agree we shouldn't have a requirement that device code support these types, but the other side of that coin is the question of how are users supposed to know?

For now, we are relying on vendor extensions for this.  If a vendor wants to support these (or any) `std` types in device code, they can create an extension for that.  We have an established way for application code to test whether an extension is supported (via macros), so applications would use that to determine if it is safe to use these types in device code.

In SYCL-Next, I think we should consider adding some guarantee to the core language saying that certain `std` types can be used in device code.

Reply:
The problem is more an implementation problem. A `std::` thing might be completely valid, except you might not have all the C++ standard library digested by or implemented for the device compiler...
An advantage for the SSCP model cherished by @illuhad. (We have SMCP but not SSCP in the SYCL glossary. :-( No definition in the spec itself either).

Reply:
What are the requirements on user defined trivially copyable types?  I don't see how there can be a practical distinction between those and standard library provided trivially copyable types.

Reply:
The requirements for a trivially copyable type are the same whether it is user-defined or part of the `std` library.

However, we need to keep in mind that "trivially copyable" is not a sufficient requirement for the type to be used in device code.  Device code has other requirements, such as a prohibition on exceptions and recursion.  Therefore, even if we agree that the types listed in this section are trivially copyable, there is still no guarantee that they can be used in device code.